### ISSUE TITLE
Drop matrix expressions support in 5.x branch

### ISSUE THREAD

**Issue by asmorkalov** [opencv/opencv#24908]
### Describe the feature and motivation

Matrix expressions API works like syntax sugar for Matrices, Scalars and some other types and allows to define complex mathematical expressions  with simple operators like +-*/. The solution also tries to optimize computation graph. Arguments:
- **ambiguousness of the expressions**. There is no way to define how Scalar should be interpreted. It could be used as single value for all channels or as (x, 0,0,0) and applied for different channels only. Broadcasting is not included in design. Related tickets and PRs:  https://github.com/opencv/opencv/issues/23978, https://github.com/opencv/opencv/issues/16538, https://github.com/opencv/opencv/pull/16597.
- **dependencies and binary size**. The expressions engine has strong dependency with all arithmetic operations for all data types. It forces linked to add 1+ mb of dead code to binary in case of static linkage.
- Most of basic operators are already overloaded for cv::Mat, cv::Rect, cv::Vec<> and  the removal will affect not so many apps and libraries. The expressions are relevant for C++ only. Python, Java and other bindings are not affected

### Additional context

The removal should happen before 5.0 release to presume compatibility in future.

**Comment by mshabunin** [comment]
To elaborate on the _binary size_ point a little bit:

* build static OpenCV with reduced dependencies:
   ```
   cmake \
    -DBUILD_JAVA=OFF \
    -DBUILD_opencv_python3=OFF \
    -DBUILD_SHARED_LIBS=OFF \
    -DCPU_DISPATCH= \
    -DWITH_OPENCL=OFF \
    -DWITH_EIGEN=OFF \
    -DWITH_IPP=OFF \
    ../opencv
   ```
* build small app (add linker options for reduced size: `target_link_options(an_app PRIVATE "-Wl,--gc-sections")`)
   ```
   #include "opencv2/core.hpp"
   int main()
   {
       cv::Mat m(100, 100, CV_8UC3, cv::Scalar(0));
       m = m + m;
       return 0;
   }
   ```
* file size is 1.17 MiB, it includes large functions such as `cv::cpu_baseline::MulTransposedR`, `cv::compare`, `cv::Mat::convertTo`, `cv::reshape`, etc. - not related to actual operation being performed.

##### Experiment 1:
* replace `m = m + m` -> `cv::add(m, m, m)` and rebuild the app with the same options
* file size is 512 KiB, it includes less unrelated functions

##### Experiment 2:
* remove `m = m + m` and rebuild the app with the same options
* file size is 288 KiB
* actual function being used for addition weighs about 9-10 KiB, so, in theory, in we should be geting ~300-350 KiB binary for simple addition operation

**Note:** I also tried to rebuild OpenCV with enabled Link Time Optimization (LTO): `-DENABLE_LTO=ON`, but for some reason I did not get expected size reduction with it :disappointed: 

**Note:** Ubuntu 22.04, GCC 11.4.0, OpenCV @ 4.x

**Comment by vpisarev** [comment]
The code example is rather artificial. It generates zero matrix, not reads any images, does not read any file storage. How realistic is it?

Matrix expressions should rather be redesigned than dropped completely. A few options are possible.
Note that there are a few other functionality blocks in OpenCV, like Imgcodecs, G-API and DNN that imply linking of a big amount of code even when it's not actually used. OpenCL (even though it was turned off here) also automatically includes a lot of code even when you don't use it. 

I believe, code size is the lowest-priority thing that shall be considered in the case of OpenCV.

**Comment by mshabunin** [comment]
Preliminary analysis has shown that expressions are widely used in OpenCV source code:
* core - 18 matches
* imgproc - 52 matches
* calib3d - 652 matches
* objdetect - 39 matches
* dnn - 49 matches

Notes:
* checked using clang-query tool:
   ```  
   clang-query-14 -p <build> <src>/opencv/modules/<module>/src/**/*.cpp
   ```
   query: 
   ```
   match callExpr(
     hasDeclaration(
       functionDecl(
         returns(
           hasDeclaration(
             namedDecl(
               hasName(
                 "MatExpr")))))), 
     isExpansionInMainFile())
   ```
* only counted functions returning MatExpr, there are also functions accepting MatExpr as an argument, template functions (?)
* checked _<module>/src/**/*.cpp_ contents only (no headers, tests, samples, apps, etc.)
* x86_64 Linux build / default configuration (might affect overall coverage)

I'll continue investigations.

**Comment by mshabunin** [comment]
I've found one border case where current expression implementation works slightly different than might be expected - https://github.com/opencv/opencv/blob/76548e29bd62f299386cb919824b34ce4baf73cb/modules/core/test/test_operations.cpp#L249-L250 these cases can be boiled down to:
```.cpp
Mat fours(3, 3, CV_16SC1, Scalar::all(4));
Mat fives(3, 3, CV_16SC1, Scalar::all(5));
Mat res = fours + 0.5 + 0.5;
```
current behavior with MatExpr: `res == fives` (`0.5 + 0.5` is calculated before adding to Mat)
if we replace operators to return Mat: `res == fours` (operators are calculated in expected order)

**Comment by mshabunin** [comment]
Another issue I have found after rewriting operators to return plain `cv::Mat` is as follows. 
Consider the following function written in a typical manner - we have an output argument and call `OutputArray::getMat()` to acquire real object to write into:
```.cpp
void fun(..., OutputArray dst)
{
    dst.create(...);
    Mat dst_ = dst.getMat();
    ...
    Mat a = ...;
    Mat b = ...;
    dst_ = a & b;
}
```
Thing is `dst_` here is actually a copy of original `Mat` object passed to `fun`, they share `data` buffer, but other fields are separate. Currently, `dst_ = a & b` is being transformed to something like `cv::and(a, b, dst_)` by the magic of expressions. But if we would return `Mat` from `operator &` here, then it will just replace contents of `dst_` object (`operator =`) and **will not affect object passed in the `dst` argument**.

Possibly it could be solved if `getMat()` returned reference to underlying object - `Mat &`, but it won't fix cases when underlying object is `std::vector` for example.

Thus, as I understand we can not simply make plain overloaded operators to replace matrix expressions without breaking current codebase. We still would need some kind of `MatExpr` which is able to fold at least simple expressions like `a = b op c`.

**Comment by pfeatherstone** [comment]
Am I understanding this right. You want to remove expression templates? That's what makes `cv::Mat` so powerful and ultimately generates performant linear algebra code.

**Comment by asmorkalov** [comment]
It's under discussion for now. The current implementation has several design bugs and produce unexpected result.

**Comment by pfeatherstone** [comment]
> It's under discussion for now. The current implementation has several design bugs and produce unexpected result.

Is that because of the typical gotchas with expression templates one has to take of when writing an implementation? For example, detecting aliasing, introducing temporaries when required (to avoid aliasing for example), data alignment, etc. I'm not sure about Eigen, but I know for sure the dlib library does a really good job of dealing with all these subtleties automatically and the user is blissfully unaware.  

Is it a case they need a fresh implementation rather than completely dropping support for this?

**Comment by pfeatherstone** [comment]
> The current implementation has several design bugs and produce unexpected result.

Surely adding more unit tests in the CI would help track these bugs and help with the fixes? Maybe that's already the case. Forgive me if so. 

Personally, the expression template engine is one of the core reasons I use `cv::Mat` and opencv in general.
